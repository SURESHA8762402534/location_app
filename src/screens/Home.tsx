import React, { useEffect, useRef } from "react";
import { View, StyleSheet, Text, Alert } from "react-native";
import moment from "moment";

import { useAppState } from "../hooks/useAppState";
import { useDispatch } from "../hooks/useDispatch";
import { useInterval } from "../hooks/useInterval";
import { getLocation, getNewLocationHistory } from "../utils/helpers";
import { getLocationDetails, checkHttpStat } from "../actions";
import { SET_LOCATION_HISTORY, SET_CURRENT_LOCATION } from "../contexts/actions";
import { State } from "../types";

import CurrentLocation from "../components/CurrentLocation";
import PreviousLocations from "../components/PreviousLocations";
import Button from "../components/Button";

const Home: React.FC = () => {

  const state = useAppState();
  const dispatch = useDispatch();
  const stateRef = useRef<State>(state);

  const saveCurrentLocation = async () => {
    const data = await getLocation();
    if(data){
      const coords = {
        latitude: data.coords.latitude,
        longitude: data.coords.longitude
      }
      const locationDetails = await getLocationDetails(coords.latitude, coords.longitude);
      if(locationDetails.success){
        const statusData = await checkHttpStat(locationDetails.data);
        if(statusData.success){
          const newHistory = getNewLocationHistory(
            stateRef.current?.current,
            stateRef.current?.history
          );
          dispatch({
            type: SET_CURRENT_LOCATION,
            payload: {
              current: {
                name: locationDetails.data,
                time: moment().format("DD/MM/YYYY, hh:mm:ss"),
                coords: coords
              },
              history: newHistory
            }
          });
        } else {
          Alert.alert("Error", statusData.data);
        }
      } else {
        Alert.alert("Error", locationDetails.data);
      }
    }
  }

  const removeItem = (index: number) => {
    const newHistory = state?.history.filter((_, idx) => idx !== index);
    dispatch({
      type: SET_LOCATION_HISTORY,
      payload: newHistory
    });
  }

  const clearAll = () => {
    dispatch({
      type: SET_LOCATION_HISTORY,
      payload: []
    });
  }
  
  useInterval(() => {
    saveCurrentLocation();
  });

  useEffect(() => {
    stateRef.current = state;
  }, [state.current.time, state.history.length]);

  return (
    <View style={styles.container}>
      <Text style={styles.header}>Location Manager</Text>
      <Text style={styles.subHeader}>Current Location</Text>
      <CurrentLocation data={state.current} />
      <Text style={styles.subHeader}>Previous Locations</Text>
      <PreviousLocations history={state.history} removeItem={removeItem} />
      <Button
        title="Clear All"
        onPress={clearAll}
        style={styles.button}
        textStyle={styles.buttonText}
      />
    </View>
  );
}

const styles = StyleSheet.create({
  container: {
    flex: 1,
    padding: "5%",
    position: "relative"
  },
  header: {
    color: "#333",
    fontSize: 20,
    fontWeight: "bold",
    marginBottom: 10
  },
  subHeader: {
    color: "#999",
    fontSize: 16,
    marginBottom: 8
  },
  button: {
    position: "absolute",
    bottom: "2%",
    right: "5%",
    width: "100%",
    backgroundColor: "blue",
    paddingVertical: 15
  },
  buttonText: {
    color: "#fff",
    fontWeight: "bold"
  }
});

export default Home;