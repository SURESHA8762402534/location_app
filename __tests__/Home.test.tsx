import { render } from "@testing-library/react-native";

import Home from "../src/screens/Home";

describe("Home screen", () => {
  it("test", () => {
    const check = true;
    expect(check).toBe(true);
  });

  it("renders properly", () => {
    render(<Home />);
  });

  it("matches snapshot", () => {
    const tree = render(<Home />);
    expect(tree.toJSON()).toMatchSnapshot();
  });

  it("has 6 children", () => {
    const tree = render(<Home />);
    expect(tree.toJSON().children.length).toBe(6);
  });

  it("has matches testID", () => {
    const tree = render(<Home />);
    expect(tree.findByTestId("test-id")).toBeTruthy();
  });
});