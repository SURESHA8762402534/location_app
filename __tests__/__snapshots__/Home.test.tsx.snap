// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Home screen matches snapshot 1`] = `
<View
  style={
    Object {
      "flex": 1,
      "padding": "5%",
      "position": "relative",
    }
  }
>
  <Text
    style={
      Object {
        "color": "#333",
        "fontSize": 20,
        "fontWeight": "bold",
        "marginBottom": 10,
      }
    }
  >
    Location Manager
  </Text>
  <Text
    style={
      Object {
        "color": "#999",
        "fontSize": 16,
        "marginBottom": 8,
      }
    }
  >
    Current Location
  </Text>
  <View
    style={
      Object {
        "flexDirection": "row",
        "marginBottom": 10,
      }
    }
  >
    <View
      style={
        Object {
          "alignItems": "center",
          "backgroundColor": "orange",
          "borderRadius": 25,
          "height": 50,
          "justifyContent": "center",
          "marginRight": 10,
          "width": 50,
        }
      }
    >
      <Text
        style={
          Object {
            "color": "#fff",
            "fontSize": 20,
            "fontWeight": "bold",
            "textTransform": "uppercase",
          }
        }
      >
        NA
        
      </Text>
    </View>
    <View
      style={
        Object {
          "flex": 1,
        }
      }
    >
      <Text
        numberOfLines={1}
        style={
          Object {
            "color": "#333",
            "fontSize": 20,
            "fontWeight": "bold",
            "marginBottom": 5,
            "textTransform": "capitalize",
          }
        }
      >
        
      </Text>
      <Text
        style={
          Object {
            "color": "#999",
            "fontSize": 16,
          }
        }
      >
        
      </Text>
    </View>
  </View>
  <Text
    style={
      Object {
        "color": "#999",
        "fontSize": 16,
        "marginBottom": 8,
      }
    }
  >
    Previous Locations
  </Text>
  <RCTScrollView
    data={Array []}
    getItem={[Function]}
    getItemCount={[Function]}
    keyExtractor={[Function]}
    onContentSizeChange={[Function]}
    onLayout={[Function]}
    onMomentumScrollBegin={[Function]}
    onMomentumScrollEnd={[Function]}
    onScroll={[Function]}
    onScrollBeginDrag={[Function]}
    onScrollEndDrag={[Function]}
    removeClippedSubviews={false}
    renderItem={[Function]}
    scrollEventThrottle={50}
    stickyHeaderIndices={Array []}
    viewabilityConfigCallbackPairs={Array []}
  >
    <View />
  </RCTScrollView>
  <View
    accessible={true}
    collapsable={false}
    focusable={true}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      Object {
        "backgroundColor": "blue",
        "borderRadius": 8,
        "bottom": "2%",
        "opacity": 1,
        "paddingHorizontal": 20,
        "paddingVertical": 15,
        "position": "absolute",
        "right": "5%",
        "width": "100%",
      }
    }
  >
    <Text
      style={
        Array [
          Object {
            "color": "#777",
            "fontSize": 16,
            "textAlign": "center",
          },
          Object {
            "color": "#fff",
            "fontWeight": "bold",
          },
        ]
      }
    >
      Clear all
    </Text>
  </View>
</View>
`;
